driver-core/src/main/java/com/datastax/driver/core/RegularStatement.java

```diff
@@ -190,6 +190,44 @@ public boolean hasValues() {
     }
 
     /**
+     * Returns the keyspace this query operates on.
+     * <p/>
+     * Unless the keyspace has been explicitly set through {@link #setKeyspace},
+     * this method will return {@code null} to avoid having to parse the query
+     * string.
+     *
+     * @return the keyspace set through {@link #setKeyspace} if such keyspace was
+     * set, {@code null} otherwise.
+     * @see Statement#getKeyspace
+     */
+    @Override
+    public String getKeyspace() {
+        return null;
+    }
+
+    /**
+     * Sets the keyspace this query operates on.
+     * <p/>
+     * This method allows you to manually provide a keyspace for this query.  It is used for the following:
+     * <ol>
+     * <li>To indicate to cassandra what keyspace the statement is applicable to (protocol V5+ only).  This is useful
+     * when the query does not provide an explicit keyspace and you want to override the session's keyspace.</li>
+     * <li>By {@link com.datastax.driver.core.policies.TokenAwarePolicy}</li> to help identify which
+     * replicas are applicable to send this statement to.</li>
+     * </ol>
+     * Do note that if the query does not use a fully qualified keyspace, then
+     * you do not need to set the keyspace through this method as the
+     * currently logged in keyspace will be used if it is set.
+     *
+     * @param keyspace the name of the keyspace this query operates on.
+     * @return this {@code SimpleStatement} object.
+     * @see Statement#getKeyspace
+     */
+    public RegularStatement setKeyspace(String keyspace) {
+        throw new UnsupportedOperationException("No concrete implementation of setKeyspace defined for " + this.getClass().getName());
```

Comment:
i had kind of waffled back and forth on that.  I ultimately did this to avoid adding a clirr exception, but I do like the idea of making it abstract to force implementation.  Will change it.

Reply:
To be clear, I think we must avoid the Clirr exception. My suggestion was to not declare it at all in the abstract class, not make it abstract.

Reply:
I see what you are saying now, that makes complete sense to me.  `setKeyspace` is not really something used within the driver itself (afaict), so there's really no need to be part of `RegularStatement`.  Instead it is only used by client APIs, who have a more concrete implementation, like `BuiltStatement` for example.